crates/neon-macros/src/export/function/mod.rs

```diff
@@ -130,53 +145,244 @@ pub(super) fn export(meta: meta::Meta, input: syn::ItemFn) -> proc_macro::TokenS
     .into()
 }
 
-// Get the ident for the first argument
-fn first_arg_ty(sig: &syn::Signature) -> Option<&syn::Ident> {
-    let arg = sig.inputs.first()?;
-    let ty = match arg {
-        syn::FnArg::Receiver(v) => &*v.ty,
-        syn::FnArg::Typed(v) => &*v.ty,
-    };
+// Determine the number of arguments to the function
+fn count_args(sig: &syn::Signature, has_context: bool, has_this: bool) -> usize {
+    let n = sig.inputs.len();
 
-    let ty = match ty {
-        syn::Type::Reference(ty) => &*ty.elem,
-        _ => return None,
+    match (has_context, has_this) {
+        (true, true) => n - 2,
+        (false, false) => n,
+        _ => n - 1,
+    }
+}
+
+// Generate the context extraction and argument for the function
+fn context_parse(
```

Comment:
Some of this parsing and checking logic feels a bit intermingled. That might just be a side effect of the code review UI, not sure.

A couple ideas you could consider for modularizing it a bit:

- Separating parsing logic into a helper module.
- Make the parsers just return an error type with the data needed for the validation, and separate logic for taking the error information and producing the syntax error. Not sure if this doesn't end up changing much.

Just some ideas.

Reply:
I hear you, but I'm concerned that separating the logic _feels_ better but actually makes the code more difficult to understand and change.

These methods only have a *two* correct branches:
* Optional argument not provided
* Optional argument present

All the other branches are for custom error messages on common mistakes. If we separate the logic, we end up with a bunch of unique variants that only serve to map to an error message. Adding or improving error messages becomes a three step process of add enum variant, parse enum variant, match enum variant instead of a single step.